packages/@uppy/core/src/Uppy.ts

```diff
@@ -918,13 +918,18 @@ export class Uppy<
       this.#restricter.getMissingRequiredMetaFields(file)
 
     if (missingFields.length > 0) {
-      this.setFileState(file.id, { missingRequiredMetaFields: missingFields })
+      this.setFileState(file.id, {
+        missingRequiredMetaFields: missingFields,
+        error: error.message,
+      })
```

Comment:
This is the simplest approach I could come up with: setting error in the file state in case of missing fields, which would make it retryable, and adding a check in `#getFilesToRetry()` for `missingRequiredMetaFields`.

This is still hacky though.